flow-report/src/summary/opportunities.tsx

```diff
@@ -0,0 +1,73 @@
+/**
+ * @license Copyright 2021 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+
+import {FunctionComponent} from 'preact';
+import {useMemo} from 'preact/hooks';
+
+import {Util} from '../../../report/renderer/util';
+import {useFlowResult} from '../util';
+
+const MAX_TOP_OPPORTUNITIES = 5;
+
+function computeTopOpportunities(reportResults: LH.ReportResult[], categoryId: string) {
```

Comment:
I think we have a few possible goals here:

- Surface the absolute most important things to fix across the entire flow
- Surface issues that are actionable and/or easy to fix
- Surface issues that are very common/high ROI 

My initial take is that the first is rather difficult to accomplish for us, and possibly not actionable for the developer either (i.e. if you got a report that just listed the "Top Audits" as the 6 metrics, that wouldn't be very interesting). I'm inclined to favor the "very common"/"high ROI" approach for non-perf categories and the "actionable"+"impact" approach for perf category. What this means in practice for a proposed algorithm...

Non-perf
1. Group all failures across reports by audit ID
2. Map each group to compute the sum of number of details items 
3. Return the top N audits when sorted by # of total details items

Perf
1. Group all opportunities across reports by audit ID
2. Map each group to compute the sum of overallSavingsMs.
3. Return the top N audits when sorted by # of total ms savings.
4. If no items are returned, fallback to non-perf algorithm.

It looks like what you have here is pretty close to non-perf algo, just using the # of reports it appears in rather than number of details items, and using weight as a tiebreaker. Is that right? WDYT about the above?

Reply:
> It looks like what you have here is pretty close to non-perf algo, just using the # of reports it appears in rather than number of details items, and using weight as a tiebreaker. Is that right? WDYT about the above?

Yup that's the algo

>  if you got a report that just listed the "Top Audits" as the 6 metrics, that wouldn't be very interesting

WDYT about filtering out any audits under group "metrics"?

> Non-perf
> 
> Group all failures across reports by audit ID
> Map each group to compute the sum of number of details items
> Return the top N audits when sorted by # of total details items

I'm worried about using the details items because audits like `viewport` don't include any details items. I understand that using # of relevant reports isn't perfect either, but it will be easier to rank all the audits including ones without items. Maybe we could sort by # reports, then # details items, then weight?

> Perf
> 
> Group all opportunities across reports by audit ID
> Map each group to compute the sum of overallSavingsMs.
> Return the top N audits when sorted by # of total ms savings.
> If no items are returned, fallback to non-perf algorithm.

Ranking first by savings SGTM, but I would still want to include some diagnostics with 0 savings if there is available space.

Reply:
> WDYT about filtering out any audits under group "metrics"?

SGTM

> I'm worried about using the details items because audits like viewport don't include any details items. I understand that using # of relevant reports isn't perfect either, but it will be easier to rank all the audits including ones without items. Maybe we could sort by # reports, then # details items, then weight?

WFM

> Ranking first by savings SGTM, but I would still want to include some diagnostics with 0 savings if there is available space.

Yes agreed. 


To solve multiple of these concerns, what about # of reports weighted by score? i.e. we wouldn't want to surface a yellow opportunity above a red diagnostic.

Reply:
> To solve multiple of these concerns, what about # of reports weighted by score?

SGTM